### PR TITLE
rukpakctl scaffolding

### LINKED ISSUE #369: Add initial scaffolding for rukpakctl
`rukpakctl` is a new CLI tool that interacts with rukpak resources. See [the design doc](https://hackmd.io/AJ8ygfzbTrmXv_CIiUCPtQ?view) for a more detailed document outlining the tool. This issue is for adding the baseline code necessary to add the binary to the project. 

A/C:
* Add new folder and main.go under /cmd
* Bootstrap a cobra CLI tool in the new folder
* Add Makefile targets

### PR DESCRIPTION
**Pull request by akihikokuroda**
Signed-off-by: akihikokuroda <akuroda@us.ibm.com>

Closes #369 

This has just scaffolding for the rukpakctl.  Are there anything to be added in this PR?

### CODE REVIEW

**Review by exdx — COMMENTED** [review]
Thanks for putting together the scaffolding. I know from the [design doc](https://hackmd.io/AJ8ygfzbTrmXv_CIiUCPtQ) we had quite a number of possible commands, but I think it would make sense to just add one at a time. 

The two commands that seem the most plausible were one that unpacks the tar.gz bundle from the volume and prints the contents as a YAML stream, and a command that could create a BI on cluster from a local set of manifests (this probably requires #92 though). Since that's the case, can we simply drop all the other commands for now and just introduce the `unpack` command for now? It would effectively do what's described in the [plain provisioner script](https://github.com/operator-framework/rukpak/blob/main/internal/provisioner/plain/README.md#make-bundle-content-available-but-do-not-install-it). It might make more sense to bring in the unpack command as a follow-up though. 

My thinking with this PR is we can have just a minimal stub for the binary, so we can develop commands iteratively afterwards. Just a main.go and some Makefile targets with the basic directory structure is enough.

File: Makefile [hunk 1]
```diff
@@ -216,3 +221,5 @@ $(GORELEASER): $(TOOLS_DIR)/go.mod # Build goreleaser from tools folder.
 	cd $(TOOLS_DIR); go build -tags=tools -o $(BIN_DIR)/goreleaser github.com/goreleaser/goreleaser
 $(KIND): $(TOOLS_DIR)/go.mod
 	cd $(TOOLS_DIR); go build -tags=tools -o $(BIN_DIR)/kind sigs.k8s.io/kind
+$(COBRA_CLI): $(TOOLS_DIR)/go.mod
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply Makefile:224]: OK.  I'll take it out.

File: cmd/rukpakctl/cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func init() {
+	cobra.OnInitialize(initConfig)
+
+	// Here you will define your flags and configuration settings.
+	// Cobra supports persistent flags, which, if defined here,
+	// will be global for your application.
+
+	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.rukpakctl.yaml)")
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply cmd/rukpakctl/cmd/root.go:59]: OK.  I'll take this out, too.

**Comment by akihikokuroda** [comment]
Thanks for review.  I'll take out unnecessary pieces for now.  I'll just put the `unpack` command stub.

File: cmd/rukpakctl/main.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/*
+Copyright © 2022 NAME HERE <EMAIL ADDRESS>
+
+*/
```
Comment (timflannagan) [inline cmd/rukpakctl/main.go:4]: nit: we can drop this copyright header?

File: cmd/rukpakctl/cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+package cmd
+
+import (
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
```
Comment (timflannagan) [inline cmd/rukpakctl/cmd/root.go:14]: I think my preference would be to have a constructor that returns a *cobra.Command:

```go
func NewRootCmd() {
    return &cobra.Command {
        ...
    }
```

File: cmd/rukpakctl/cmd/root.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	// has an action associated with it:
+	// Run: func(cmd *cobra.Command, args []string) { },
+}
+
+// Execute adds all child commands to the root command and sets flags appropriately.
+// This is called by main.main(). It only needs to happen once to the rootCmd.
+func Execute() {
+	err := rootCmd.Execute()
+	if err != nil {
+		os.Exit(1)
```
Comment (timflannagan) [inline cmd/rukpakctl/cmd/root.go:33]:
```suggestion
                fmt.Println(os.Stderr, err)
		os.Exit(1)
```

File: cmd/rukpakctl/cmd/root.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:   "rukpakctl",
+	Short: "A brief description of your application",
+	Long: `A longer description that spans multiple lines and likely contains
+examples and usage of using your application. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
```
Comment (timflannagan) [inline cmd/rukpakctl/cmd/root.go:22]: nit: Can we populate these short/long descriptions?

File: cmd/rukpakctl/main.go [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+package main
+
+import "github.com/operator-framework/rukpak/cmd/rukpakctl/cmd"
+
+func main() {
+	cmd.Execute()
```
Comment (timflannagan) [inline cmd/rukpakctl/main.go:10]: If we refactor the root cobra.Command into a constructor function vs. relying on a package-level variable, we can do the following:

```suggestion
	cmd := root.NewCmd()
	if err := cmd.Execute(); err != nil {
	    fmt.Fprintln(os.Stderr, err)
	    os.Exit(1)
        }
```

File: cmd/rukpakctl/cmd/root.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/*
+Copyright © 2022 NAME HERE <EMAIL ADDRESS>
+
+*/
+package cmd
```
Comment (timflannagan) [inline cmd/rukpakctl/cmd/root.go:5]:
```suggestion
package root
```

File: cmd/rukpakctl/cmd/unpack.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+package cmd
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+)
+
+// unpackCmd represents the unpack command
+var unpackCmd = &cobra.Command{
```
Comment (timflannagan) [inline cmd/rukpakctl/cmd/unpack.go:14]: I think we can either drop this unpack command, or apply the same feedback where we can use a constructor that returns a *cobra.Command, which the root.NewCmd constructor can register:

```go
package root

import (
    ...
)

func NewCmd() *cobra.Command {
    cmd := &cobra.Command{
        ...
    }

    cmd.AddCommand(unpack.NewCmd())
    ...
    return cmd
}
```

**Review by timflannagan — COMMENTED** [review]
Thanks for taking a stab at this - I had a couple of comments, but I think we can use a constructor function that returns built up cobra.Commands vs. relying on package-level variable declarations and `init` functions. Thoughts?

File: cmd/rukpakctl/cmd/root.go (hunk 4 above)
Comment (timflannagan) [reply cmd/rukpakctl/cmd/root.go:14]: Actually: what we can do here is move this root.go into it's own namespace, e.g. `package root` so the cmd/rukpakctl/main.go can call it via `root.NewCmd()`.

**Comment by akihikokuroda** [comment]
@timflannagan Thanks for review.  The contents in the `rukpakctl` are generated by 
```
cobra-cli init
cobra-cli add unpack
```
[as-is](https://github.com/spf13/cobra-cli/blob/main/README.md).  I thought it was intention of [this issue](https://github.com/operator-framework/rukpak/issues/369).  I can take your comments and make changes if it should be done in this issue.  WDYT?

**Comment by timflannagan** [comment]
I'm fine with a follow up here. It looks like Dan already approved so feel free to merge.

**Pull request merged**